Formalize the Supervisor's HTTP API

raskchanky opened the pull request:
This work is analogous to the work done in https://github.com/habitat-sh/habitat/pull/4823 to formalize the supervisor's rendering context, but for the HTTP API instead.

The issue that this PR closes contains all of the gory details, but the short of it is, we need to have more control over the output of the HTTP API endpoints.

Currently, we're just serializing a bunch of internal data structures directly to JSON and serving that up.  That's sub-optimal for a few reasons:

* It makes it difficult to refactor the internals without breaking people that are relying on the output of the HTTP API endpoints
* It exposes implementation details to end users
* It creates a mass of duplicate data that users need to wade through to get to the data they're actually interested in

This PR introduces 3 proxy data structures that more or less wrap our internal ones for now, but give us the facade we need to be able to refactor the internals without breaking out external contract.

Additionally, 3 JSON schemas have been introduced that describe the `/butterfly`, `/census` and `/services` endpoints and tests have been written that verify that the output of these endpoints conform to the schema.

Closes https://github.com/habitat-sh/habitat/issues/5389

![](https://media.giphy.com/media/rIQRpLWc7jVbq/giphy.gif)

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

christophermaier commented on components/sup/doc/http_gateway_butterfly_schema.json:
```diff
@@ -0,0 +1,483 @@
+                  },
+                  "status": {
+                    "type": "string",
+                    "description": "The status of the election"
```
We should enumerate the possible values here. The more information we can provide users, the better, particularly since we can also automatically generate user-facing documentation from this schema.

raskchanky replied:
Will do

christophermaier commented on components/sup/doc/http_gateway_butterfly_schema.json:
```diff
@@ -0,0 +1,483 @@
+          "description": "The health of the members",
+          "additionalProperties": {
+            "type": "string",
+            "description": "The health of the member, e.g. Alive, Suspect, Confirmed, Departed"
```
These possible values should be formally enumerated, instead of being part of a documentation string.

raskchanky replied:
Will do

christophermaier commented on components/sup/doc/http_gateway_census_schema.json:
```diff
@@ -0,0 +1,316 @@
+                  "description": "The environment portion of a service group specification",
+                  "additionalProperties": {
+                    "type": ["null", "boolean", "string", "integer", "array", "object"]
+                  }
```
Isn't `environment` just a String?

raskchanky replied:
Yeah, you're right.  I don't know how I missed that, thanks.

christophermaier commented on components/sup/doc/http_gateway_census_schema.json:
```diff
@@ -0,0 +1,316 @@
+                  "type": "boolean",
+                  "description": "Whether this member is considered possibly unreachable, from a network perspective"
+                },
+                "sys": {
```
I wonder if there's a straightforward way for us to share type definitions across our schema files. For instance, I think this is essentially the same as https://github.com/habitat-sh/habitat/blob/d68309486b02dca9491ed4d9933e40e835631a20/components/sup/doc/render_context_schema.json#L219-L280

Of course, that raises the legitimate question of whether or not we truly want these structures to be the same between the HTTP gateway and the rendering context (maybe? :man_shrugging:)

In any event, it seems there are similar structures shared between the various HTTP gateway endpoints, so ensuring all those (at the very least) are in-sync with the same schema definitions would be nice.

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+      },
+      "binding_mode": {
+        "type": "string",
+        "description": "What type of bind this is, either strict or relaxed"
```
This should be enumerated.

raskchanky replied:
Will do

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+        }
+      },
+      "channel": {
+        "type": "string",
```
Shouldn't this be nullable? Not every service is configured to update.

raskchanky replied:
It's not optional [in the struct](https://github.com/habitat-sh/habitat/blob/master/components/sup/src/manager/service/mod.rs#L98)

christophermaier replied:
Whoa, that's interesting 🤔 

I bet we just end up running with a default value if not otherwise set, and then whether that actually does anything is governed by the update strategy.

That would be interesting to clean up at some point (not here, though).

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+      },
+      "desired_state": {
+        "type": "string",
+        "description": "The desired state for this service, either Up or Down"
```
Enumerate values.

raskchanky replied:
Sorry, I didn't even know this was a thing you could do when I wrote this.  I'll fix all these.

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+      },
+      "health_check": {
+        "type": "string",
+        "description": "The results of the last health check, either Ok, Warning, Critical or Unknown"
```
Enumerate values.

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+            "description": "The environment for this package",
+            "additionalProperties": {
+              "type": ["null", "boolean", "string", "integer", "array", "object"]
+            }
```
I don't think this is correct... `env` is an object of environment variables, so it should have String keys and values (might need to check how it works with values that are numbers, booleans, etc... do they actually get encoded as Strings?)

raskchanky replied:
You're right.  I have no idea what I was thinking.

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+            "description": "The exports for this package",
+            "additionalProperties": {
+              "type": "string"
+            }
```
Similarly, this should just be an object of String keys and values.

raskchanky replied:
Right, is that not what I have?

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+          },
+          "state": {
+            "type": "string",
+            "description": "The state of this process, either Up or Down"
```
Enumerate values.

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+          },
+          "state_entered": {
+            "type": "integer",
+            "description": "The time the process entered its current state"
```
I can never remember this... does this get rendered as seconds from the epoch? If so, it'd be good to document that here (and probably in the code, too).

raskchanky replied:
Yes it does.  I added docs to that effect.

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+      },
+      "smoke_check": {
+        "type": "string",
+        "description": "The status of this service's smoke check, either Ok, Failed, or Pending"
```
Enumerate values.

Though really we should make a note here that smoke checks aren't actually a thing... I think the main code for them got pulled a while ago, but it was an incomplete removal. 🙄

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+      },
+      "topology": {
+        "type": "string",
+        "description": "The topology this service runs as, either Standalone or Leader"
```
Enumerate values.

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+      },
+      "update_strategy": {
+        "type": "string",
+        "description": "The strategy used to update this service, either None, AtOnce, or Rolling"
```
Enumerate values.

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -0,0 +1,521 @@
+          { "type": "null" }
+        ]
+      },
+      "sys": {
```
See earlier comments about sharing common definitions.

raskchanky commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -980,3 +983,71 @@ impl fmt::Display for Service {
+// worrying about breaking the data returned to users.
+pub struct ServiceProxy<'a> {
+    service: &'a Service,
+    render_config: bool,
```
Gotcha, will do

raskchanky commented on components/sup/src/manager/mod.rs:
```diff
@@ -1442,62 +1442,54 @@ impl Manager {
+            .map(|s| ServiceProxy::new(s, !feat::is_enabled(feat::RedactHTTP)))
+            .collect();
+
+        services_to_render.extend(watched_service_proxies);
```
Yeah, not gonna lie, this code felt _really good_ to write.

raskchanky commented on components/butterfly/src/server/mod.rs:
```diff
@@ -1017,7 +1017,7 @@ impl Serialize for Server {
         S: Serializer,
     {
-        let mut strukt = serializer.serialize_struct("butterfly", 6)?;
+        let mut strukt = serializer.serialize_struct("butterfly", 7)?;
```
It doesn't seem to break anything if you give it the wrong count, but my OCD + general pedantry doesn't allow me to let things like this slide.

christophermaier replied:
I just opted out altogether:

https://github.com/habitat-sh/habitat/blob/d68309486b02dca9491ed4d9933e40e835631a20/components/sup/src/templating/context.rs#L237-L239

(That is assuming we're only ever serializing to JSON... TOML may work too, but I forget.)

raskchanky commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -980,3 +983,71 @@ impl fmt::Display for Service {
+        let num_fields: usize = if *&self.render_config { 27 } else { 26 };
+
+        let mut strukt = serializer.serialize_struct("service", num_fields)?;
+        strukt.serialize_field("all_pkg_binds", &self.service.all_pkg_binds)?;
```
Good call

raskchanky commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -980,3 +983,71 @@ impl fmt::Display for Service {
+        }
+    }
+
+    pub fn service(&self) -> &Service {
```
It was used in a previous incarnation of this code and I forgot to remove it.  I'll delete it.

raskchanky commented on components/butterfly/src/server/mod.rs:
```diff
@@ -1056,6 +1056,35 @@ fn persist_loop(server: Server) {
+// This is a proxy struct to represent what information we're writing to the dat file, and
+// therefore what information gets sent out via the HTTP API. Right now, we're just wrapping the
+// actual Server struct, but this will give us something we can refactor against without
+// worrying about breaking the data returned to users.
```
Will do

raskchanky commented on components/sup/src/census.rs:
```diff
@@ -214,6 +216,46 @@ impl CensusRing {
+// This is a proxy struct to represent what information we're writing to the dat file, and
+// therefore what information gets sent out via the HTTP API. Right now, we're just wrapping the
+// actual CensusRing struct, but this will give us something we can refactor against without
+// worrying about breaking the data returned to users.
```
Will do (I did it on `ServiceProxy` as well)

raskchanky commented:
@christophermaier I believe I've addressed all of your comments.  I included a bit of refactoring of the rendering context schema you wrote, for the `sys` key.  Incidentally, there ended up being two different definitions for that, since we have two different versions of it in various places.

christophermaier commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -1003,20 +1003,26 @@ impl fmt::Display for Service {
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ConfigRendering {
```
Adding some documentation describing what each of these does would be great, and is one of the benefits of using Enums-as-Booleans; it gives you a really nice place to hang documentation off of.

christophermaier replied:
Also, since the intention of this enum is to act like a "named boolean", we could make it behave even more like a boolean by deriving `Copy`, which means you wouldn't actually need to derive `Clone`, and wouldn't need to call `.clone()` everywhere you pass this around, which makes for an easier-to-use API.

Clearly calling that out in the documentation string is also a good idea, because implementing `Copy` is a Big Deal (TM). We don't want to do any kind of refactoring of this type in the future that adds some kind of extra large-memory member, because then we'd immediately be copying it everywhere. That's _highly_ unlikely in this usecase, but making that very clear can't hurt.

raskchanky replied:
Ah, good call.  The `.clone()` thing in 2 places was bugging me, but I wasn't entirely sure how to get around it.

christophermaier commented on components/sup/doc/http_gateway_census_schema.json:
```diff
@@ -105,42 +95,14 @@
-                    "version"
-                  ],
-                  "type": "object"
+                  "$ref": "render_context_schema.json#/definitions/package_identifier"
```
![tenor-80977228](https://user-images.githubusercontent.com/207178/46369120-eadf0300-c64f-11e8-892b-54bf85a52609.gif)

christophermaier commented on components/sup/doc/http_gateway_butterfly_schema.json:
```diff
@@ -59,7 +114,7 @@
                   "status": {
                     "description": "The status of the election",
-                    "type": "string"
+                    "enum": ["Running", "NoQuorum", "Finished"]
```
At some point, I'd like to revisit the `Display` implementations of this enum, as well as that of health below... these are upper-case, whereas some of the other enums below are lowercase. Nothing that needs to be fixed right now, but something to keep in mind for #5692.

raskchanky replied:
I'll make a note of it on that issue.

christophermaier commented on components/sup/doc/http_gateway_services_schema.json:
```diff
@@ -19,30 +19,71 @@
-    "package_identifier": {
-      "description": "A Habitat package",
+    "sys_extended": {
+      "description": "System information",
```
Adding a `$comment` key (the `$` is important) here explaining the difference between `sys_basic` and `sys_extended` would be useful to help guide #5692.

raskchanky replied:
Will do

raskchanky commented:
Appveyor tests continue to fail, even after nuking the Appveyor cache.  The failure is one that Travis used to have, but nuking Travis's cache caused Travis tests to pass.  These same tests pass locally.  I'm inclined to merge this unless there are objections.
